src/main/java/org/testng/GuiceHelper.java

```diff
@@ -71,61 +75,61 @@ private static Module getParentModule(ITestContext context) {
     if (!Module.class.isAssignableFrom(parentModule)) {
       throw new TestNGException("Provided class is not a Guice module: " + parentModule.getName());
     }
+    if (!context.getGuiceModules((Class<? extends Module>) parentModule).isEmpty()) {
+      return context.getGuiceModules((Class<? extends Module>) parentModule).get(0);
+    }
+    Module obj;
     try {
       Constructor<?> moduleConstructor = parentModule.getDeclaredConstructor(ITestContext.class);
-      return (Module)InstanceCreator.newInstance(moduleConstructor, context);
+      obj = (Module) InstanceCreator.newInstance(moduleConstructor, context);
     } catch (NoSuchMethodException e) {
-      return (Module)InstanceCreator.newInstance(parentModule);
+      obj = (Module) InstanceCreator.newInstance(parentModule);
     }
-  }
-
-  /**
-   * @param context The test context
-   * @param moduleInstances The modules
-   * @return The Injector
-   * @deprecated - This method stands deprecated as of 7.0.1
-   */
-  @Deprecated
-  public static Injector createInjector(ITestContext context, List<Module> moduleInstances) {
-    return createInjector(context, com.google.inject.Guice::createInjector, moduleInstances);
+      context.addGuiceModule(obj);
+    return obj;
   }
 
   public static Injector createInjector(ITestContext context,
       IInjectorFactory injectorFactory, List<Module> moduleInstances) {
     Module parentModule = getParentModule(context);
+    BiPredicate<Module, Module> condition = (m,n) -> m.getClass().equals(n.getClass());
     List<Module> fullModules = Lists.newArrayList(moduleInstances);
     if (parentModule != null) {
-      fullModules.add(parentModule);
+      fullModules = Lists.merge(fullModules, Collections.singletonList(parentModule), condition);
     }
     Stage stage = Stage.DEVELOPMENT;
     String stageString = context.getSuite().getGuiceStage();
     if (isStringNotEmpty(stageString)) {
       stage = Stage.valueOf(stageString);
     }
+    fullModules.forEach(context::addGuiceModule);
     return injectorFactory.getInjector(stage, fullModules.toArray(new Module[0]));
   }
 
   private List<Module> getModules(Guice guice, Injector parentInjector, Class<?> testClass) {
     List<Module> result = Lists.newArrayList();
+    BiPredicate<Module, Module> condition = (m,n) -> m.getClass().equals(n.getClass());
     for (Class<? extends Module> moduleClass : guice.modules()) {
       List<Module> modules = context.getGuiceModules(moduleClass);
       if (modules != null && !modules.isEmpty()) {
         result.addAll(modules);
+        result = Lists.merge(result, modules, condition);
       } else {
         Module instance = parentInjector.getInstance(moduleClass);
-        result.add(instance);
-        context.getGuiceModules(moduleClass).add(instance);
+        result = Lists.merge(result, Collections.singletonList(instance), condition);
```

Comment:
What do you think about a `<T> Lists.add(List<T> list, BiPredicate<T, T> condition, E... elements)`?

Reply:
@juherr - That method was added specifically to solve this particular use case of `merge two lists using a predicate`. That is why I didnt take it to the next level. If at all there's a usecase like that, we can perhaps come back and enhance this method.